scripts/model_analysis.py

```diff
@@ -0,0 +1,887 @@
+# fail_symbol = "\u274C"        # Crossmark
+# unknown_symbol = "\uFFFD"     # Question mark
+
+# Html symbols:
+pass_symbol = "&#x2705;"  # Checkmark
+fail_symbol = "&#x274C;"  # Crossmark
+unknown_symbol = "&#xFFFD;"  # Question mark
+
+
+class CompilerComponent(IntEnum):
+    FORGE = 0
+    MLIR = 1
+    TT_METAL = 2
+    UNKNOWN = 4
+
+
+class MatchingExceptionRule:
+    """
+    Represents a rule for matching exceptions based on specific tokens.
+
+    Attributes:
+        rule_name (str): Name of the rule.
+        rule_tokens (List[str]): List of tokens to match in an exception message.
+    """
+
+    def __init__(self, rule_name: str, rule_tokens: List[str]):
+        self.rule_name = rule_name
+        self.rule_tokens = rule_tokens
+
+    def match_rule(self, exception: str):
+        """
+        Matches the rule tokens against the given exception string.
+
+        Args:
+            exception (str): Exception message to match against.
+        """
+        # Check if all tokens in rule_tokens exist in the exception message and
+        # return the rule_token if matches otherwise return None
+        matched_token = all([True if token in exception else False for token in self.rule_tokens])
+        if matched_token:
+            return " ".join(self.rule_tokens)
+        else:
+            return None
+
+
+class MatchingCompilerComponentException:
+    """
+    Represents exception matching for a specific compiler component.
+
+    Attributes:
+        compiler_component (CompilerComponent): Compiler component associated with this exception matching.
+        exception_rules (List[MatchingExceptionRule]): List of exception rules for this component.
+    """
+
+    def __init__(self, compiler_component: CompilerComponent, exception_rules: List[MatchingExceptionRule]):
+        self.compiler_component = compiler_component
+        self.exception_rules = exception_rules
+
+    def match_rule(self, exception: str):
+        """
+        Matches the given exception against the exception rules of this compiler component.
+        Args:
+            exception (str): Exception message to be checked against the rules.
+        """
+        # Iterate over all exception rules for this compiler component.
+        for rule in self.exception_rules:
+            # Attempt to match the current rule against the exception and If a match is found,
+            # return the compiler component and the constructed error message.
+            if rule.match_rule(exception) is not None:
+                match_err_msg = (
+                    self.compiler_component.name + " "
+                    if rule.rule_name is None or rule.rule_name == ""
+                    else self.compiler_component.name + ": " + rule.rule_name + " "
+                )
+                match_err_msg += rule.match_rule(exception)
+                return self.compiler_component, match_err_msg
+
+        return None, None
+
+
+common_failure_matching_rules_list = [
```

Comment:
When we start with refactor, let's put these error mappings in separated file for easier update and manage over time

Reply:
Sure Nikola I will move it to separate file while restructing

Reply:
Add the task of moving  error mappings in separated file to the restructure issues - https://github.com/tenstorrent/tt-forge-fe/issues/723